modules/analysis/src/service/load.rs

```diff
@@ -346,6 +347,267 @@ impl InnerService {
+                    DatabaseBackend::Postgres,
+                    sql,
+                    [purl.qualifier_uuid().into()],
+                );
+                let rows = connection.query_all(stmt).await?;
+                rows.into_iter()
+                    .filter_map(|row| {
+                        row.try_get_by_index::<Uuid>(0)
+                            .ok()
+                            .map(|sbom_id| sbom_id.to_string())
+                    })
+                    .collect::<Vec<String>>()
+            }
+            GraphQuery::Component(ComponentReference::Cpe(cpe)) => {
+                let sql = r#"
+                  SELECT distinct sbom_id
+                  FROM (
+                      SELECT sbom.sbom_id, sbom.published, product.id, cpe.id,
+                      RANK() OVER (PARTITION BY cpe.id ORDER BY sbom.published DESC) AS rank
+                      FROM sbom_package_cpe_ref
+                      LEFT JOIN sbom ON sbom.sbom_id = sbom_package_cpe_ref.sbom_id
+                      LEFT JOIN product_version on sbom.sbom_id = product_version.sbom_id
+                      LEFT JOIN product on product.id = product_version.product_id
+                      LEFT JOIN cpe ON sbom_package_cpe_ref.cpe_id = cpe.id
+                      WHERE sbom_package_cpe_ref.cpe_id = $1
+                  ) AS subquery
+                  WHERE rank = 1;
+                "#;
+                let stmt = Statement::from_sql_and_values(
+                    DatabaseBackend::Postgres,
+                    sql,
+                    [cpe.uuid().into()],
+                );
+                let rows = connection.query_all(stmt).await?;
+                rows.into_iter()
+                    .filter_map(|row| {
+                        row.try_get_by_index::<Uuid>(0)
+                            .ok()
+                            .map(|sbom_id| sbom_id.to_string())
+                    })
+                    .collect::<Vec<String>>()
+            }
+            GraphQuery::Query(query) => {
+                // TODO: we assume q=<partial purl> was supplied this area of the code will change.
+                //       for now creating a special local function to handle.
+                #[derive(Debug, Default, PartialEq)]
+                struct BasicPurlParts<'a> {
+                    ptype: Option<&'a str>,
+                    namespace: Option<&'a str>, // Represents the full namespace string
+                    name: Option<&'a str>,
+                    version: Option<&'a str>,
+                    qualifiers_str: Option<&'a str>, // Qualifiers as a single raw string
+                    subpath: Option<&'a str>,
+                }
+
+                /// WARNING: highly non-compliant parsing of partial pURL.
+                fn basic_non_compliant_parse_purl(purl_str: &str) -> Option<BasicPurlParts> {
+                    // start with "pkg:" if it exists
+                    let remaining = purl_str.strip_prefix("pkg:")?;
```

Comment:
This line effectively requires all `q=` values to start with "pkg:". This entirely defeats the purpose of the q parameter, and I think it'll lead to confusion. The logic prevents me from querying say `q=quarkus&purl:version>2`. 

We already have a variant for these endpoints that takes an actual purl, i.e. something that _must_ start with "pkg:". Perhaps we change the logic of that variant to have the behavior you're attempting to implement here, a kind of "loose" parsing of a purl?

Reply:
Or maybe introduce a new parameter, say "partial", that does this logic? I'd prefer we not abuse the 'q' parameter. Since it's used throughout the app, I think it's behavior should be consistent.

Reply:
as prev mentioned 'forget about all that' ... will force push and remove - once I puzzle out how to get windowing with RANK into sea_orm (or how to use filtering_with on statements).